src/main/java/gift/dao/ProductDao.java

```diff
@@ -0,0 +1,106 @@
+package gift.dao;
+
+import gift.entity.Product;
+import gift.exception.DBServerException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.simple.JdbcClient;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class ProductDao {
+    private final JdbcClient jdbcClient;
+
+    private static class ProductRowMapper implements RowMapper<Product> {
+        @Override
+        public Product mapRow(java.sql.ResultSet rs, int rowNum) throws java.sql.SQLException {
+            return new Product(
+                    rs.getLong("id"),
+                    rs.getString("name"),
+                    rs.getLong("price"),
+                    rs.getString("image_url")
+            );
+        }
+    }
+
+    public ProductDao(JdbcClient client) {;
+        this.jdbcClient = client;
+    }
+
+    public List<Product> findAll() {
+        String sql = "SELECT * FROM products";
+        return jdbcClient.sql(sql)
+                .query(new ProductRowMapper())
+                .stream()
+                .toList();
+    }
+
+    public Product findById(Long productId) {
+        String sql = "SELECT * FROM products WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(productId)
+                .query(new ProductRowMapper())
+                .stream()
+                .findFirst()
+                .orElse(null);
+    }
```

Comment:
findByID를 할 때, 결과가 없으면 orElse(null)로 반환하고있는데, Optional을 사용해서 좀 더 안전하게 처리하는 건 어떨까요?

```suggestion
public Optional<Product> findById(Long productId) {
    String sql = "SELECT * FROM products WHERE id = ?";
    return jdbcClient.sql(sql)
        .param(productId)
        .query(productRowMapper)
        .stream()
        .findFirst();
}
```